File: src/main/java/com/beat/domain/member/domain/Member.java
package com.beat.domain.member.domain;

import com.beat.domain.BaseTimeEntity;
import com.beat.domain.user.domain.Users;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.OneToOne;
import lombok.AccessLevel;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;

import org.hibernate.annotations.OnDelete;
import org.hibernate.annotations.OnDeleteAction;

import java.time.LocalDateTime;

@Entity
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class Member extends BaseTimeEntity {

	@Id
	@GeneratedValue(strategy = GenerationType.IDENTITY)
	private Long id;

	@Column(nullable = false)
	private String nickname;

	@Column(nullable = true)
	private String email;

	@Column(nullable = true)
	private LocalDateTime deletedAt;

	@OneToOne(fetch = FetchType.LAZY)
	@JoinColumn(name = "user_id", nullable = false)
	@OnDelete(action = OnDeleteAction.CASCADE)
	private Users user;

	@Column(nullable = false)
	private Long socialId;  // 카카오 회원번호 저장

	@Enumerated(EnumType.STRING)
	private SocialType socialType;

	@Builder
	private Member(String nickname, String email, LocalDateTime deletedAt, Users user, Long socialId,
		SocialType socialType) {
		this.nickname = nickname;
		this.email = email;
		this.deletedAt = deletedAt;
		this.user = user;
		this.socialId = socialId;
		this.socialType = socialType;
	}

	public static Member create(
		final String nickname,
		final String email,
		final Users user,
		final Long socialId,
		final SocialType socialType
	) {
		return Member.builder()
			.nickname(nickname)
			.email(email)
			.user(user)
			.socialId(socialId)
			.socialType(socialType)
			.build();
	}
}

File: domain/src/main/kotlin/com/beat/domain/member/model/Member.kt
package com.beat.domain.member.model

import com.beat.domain.member.vo.SocialIdentity
import com.beat.domain.sharedkernel.model.AggregateRoot
import com.beat.domain.user.model.Users
import java.time.LocalDateTime

class Member private constructor(
    private val memberId: Id?,
    val nickname: String,
    val email: String?,
    val deletedAt: LocalDateTime?,
    private val linkedUserId: Users.Id,
    val socialIdentity: SocialIdentity,
) : AggregateRoot {
    val id: Long?
        get() = memberId?.value

    val userId: Long
        get() = linkedUserId.value

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Member) return false
        return memberId != null && memberId == other.memberId
    }

    override fun hashCode(): Int = memberId?.hashCode() ?: System.identityHashCode(this)

    override fun toString(): String = "Member(id=$id)"

    @JvmInline
    value class Id private constructor(val value: Long) {
        companion object {
            fun from(value: Long): Id = Id(value)

            fun fromNullable(value: Long?): Id? = value?.let(::from)
        }
    }

    companion object {
        fun create(
            nickname: String,
            email: String?,
            userId: Long,
            socialIdentity: SocialIdentity,
        ): Member = Member(
            memberId = null,
            nickname = nickname,
            email = email,
            deletedAt = null,
            linkedUserId = Users.Id.from(userId),
            socialIdentity = socialIdentity,
        )

        fun rehydrate(
            id: Long?,
            nickname: String,
            email: String?,
            deletedAt: LocalDateTime?,
            userId: Long,
            socialIdentity: SocialIdentity,
        ): Member = Member(
            memberId = Id.fromNullable(id),
            nickname = nickname,
            email = email,
            deletedAt = deletedAt,
            linkedUserId = Users.Id.from(userId),
            socialIdentity = socialIdentity,
        )
    }
}
